### PR TITLE
Port files trashbin events to IEventDispatcher/IEventListener

### PR DESCRIPTION
**Pull request by CarlSchwan**
oc_hooks is deprecated and will at some point be removed

### CODE REVIEW

**Review by artonge — COMMENTED** [review]
Still a bunch of lint errors/warnings :)

File: apps/files_trashbin/lib/AppInfo/Application.php [hunk 1]
```diff
@@ -55,20 +62,18 @@ public function register(IRegistrationContext $context): void {
 		$context->registerServiceAlias('principalBackend', Principal::class);
 
 		$context->registerUserMigrator(TrashbinMigrator::class);
+
+		$context->registerEventListener(NodeWrittenEvent::class, EventListener::class);
+		$context->registerEventListener(BeforeUserDeletedEvent::class, EventListener::class);
+
+		// pre and post-rename, disable trash logic for the copy+unlink case
+		$context->registerEventListener(BeforeNodeDeletedEvent::class, Trashbin::class);
+		$context->registerEventListener(BeforeFileSystemSetupEvent::class, EventListener::class);
```
Comment (come-nc) [inline apps/files_trashbin/lib/AppInfo/Application.php:71]:
```suggestion

		$context->registerEventListener(NodeWrittenEvent::class, EventListener::class);
		$context->registerEventListener(BeforeUserDeletedEvent::class, EventListener::class);
		$context->registerEventListener(BeforeFileSystemSetupEvent::class, EventListener::class);

		// pre and post-rename, disable trash logic for the copy+unlink case
		$context->registerEventListener(BeforeNodeDeletedEvent::class, Trashbin::class);
```
Please reorder to make it clearer the comment is only about BeforeNodeDeletedEvent, and to group by listening class.

File: apps/files_trashbin/lib/Storage.php [hunk 2]
```diff
@@ -37,48 +37,31 @@
...
-		ITrashManager $trashManager = null,
-		IUserManager $userManager = null,
-		ILogger $logger = null,
-		EventDispatcherInterface $eventDispatcher = null,
-		IRootFolder $rootFolder = null
+		?ITrashManager $trashManager = null,
+		?IUserManager $userManager = null,
+		?LoggerInterface $logger = null,
+		?EventDispatcherInterface $eventDispatcher = null,
+		?IRootFolder $rootFolder = null
```
Comment (come-nc) [inline apps/files_trashbin/lib/Storage.php:64]: When are those null? A lot of the code is expecting non-null values.

File: apps/files_trashbin/lib/Trashbin.php [hunk 3]
```diff
@@ -44,6 +44,7 @@
  */
 namespace OCA\Files_Trashbin;
 
+use OC\Files\Node\Node;
```
Comment (come-nc) [inline apps/files_trashbin/lib/Trashbin.php:47]:
```suggestion
use OC\Files\Node;
```
psalm seems to suggest the wrong Node is imported, but not sure if this is the right fix

File: apps/files_trashbin/lib/Storage.php (hunk 2 above)
Comment (CarlSchwan) [reply apps/files_trashbin/lib/Storage.php:64]: I just adapted the old php doc types. Just to be sure to not have regressions. If we want to make them non nullable, best to do in another PR to make it easier to bisect.

File: apps/files_trashbin/lib/Trashbin.php (hunk 3 above)
Comment (CarlSchwan) [reply apps/files_trashbin/lib/Trashbin.php:47]: it wants the OCP type :)

File: apps/files_trashbin/lib/Storage.php (hunk 2 above)
Comment (come-nc) [reply apps/files_trashbin/lib/Storage.php:64]: Why is the constructor in the interface?

**Review by come-nc — APPROVED** [review]

File: apps/files_trashbin/lib/Storage.php (hunk 2 above)
Comment (CarlSchwan) [reply apps/files_trashbin/lib/Storage.php:64]: No idea :(

**Comment by tcitworld** [comment]
Rebased

File: apps/files_trashbin/lib/Trashbin.php [hunk 4]
```diff
@@ -63,22 +67,20 @@
 use OCP\Lock\LockedException;
 use Psr\Log\LoggerInterface;
 
-class Trashbin {
+class Trashbin implements IEventListener {
```
Comment (tcitworld) [inline apps/files_trashbin/lib/Trashbin.php:70]: **question:** I don't understand why it needs this interface :thinking:

File: lib/composer/composer/autoload_classmap.php [hunk 5]
```diff
@@ -954,6 +955,7 @@
     'OC\\Core\\Command\\Maintenance\\RepairShareOwnership' => $baseDir . '/core/Command/Maintenance/RepairShareOwnership.php',
     'OC\\Core\\Command\\Maintenance\\UpdateHtaccess' => $baseDir . '/core/Command/Maintenance/UpdateHtaccess.php',
     'OC\\Core\\Command\\Maintenance\\UpdateTheme' => $baseDir . '/core/Command/Maintenance/UpdateTheme.php',
+    'OC\\Core\\Command\\Preview\\Regenerate' => $baseDir . '/core/Command/Preview/Regenerate.php',
```
Comment (tcitworld) [inline lib/composer/composer/autoload_classmap.php:958]: **nitpick:** Don't know where this comes from

**Comment by skjnldsv** [comment]
As there is no feedback since a while I will close this ticket. 
If you will decide to work on this feature again and if it hasn't been fixed or implemented already, feel free to re-open and solve the various conflicts.

File: apps/files_trashbin/lib/Storage.php (hunk 2 above)
Comment (come-nc) [reply apps/files_trashbin/lib/Storage.php]: This will be dealt with in another PR as this constructor is not touched anymore by the PR after rebase.

**Comment by come-nc** [comment]
The 4 ViewTest failures came from the fact that the trashbin wrapper is now registered and intercepts the rmdir and unlink calls to applies trashbin logic instead of actually deleting stuff.
I added a call to pause trash in those tests and it passes now.

**Review by susnux — APPROVED** [review]

**Pull request merged**